File: ai/src/main/java/org/pingel/gestalt/core/FormFactory.java
package org.pingel.gestalt.core;

import java.awt.Point;
import java.awt.event.MouseEvent;

import org.pingel.gestalt.ui.Widget;


public class FormFactory
implements Widget
{
	protected Form archetype = null;

	public FormFactory()
	{
		
	}
	
    public FormFactory(Form archetype)
    {
        this.archetype = archetype;
    }


    public Form createForm()
    {
        // This is called by mousePressed
        
        Form f = archetype.duplicate();
        f.arrange(getArchetype().getCenter());
        f.setDetachable(true);

        return f;
    }

//    public Form createForm(Unifier unifier)
//    {
//        // This is called by org.pingel.causality.docalculus.*
//        
//        Form f = archetype.duplicateAndReplace();
//        f.arrange(getArchetype().getCenter());
//        f.setDetachable(true);
//    		
//        // TODO bind vars in unifier
//        return null;
//    }
    	
    public Form getArchetype()
    {
        return archetype;
    }

    public String toString()
    {
        return archetype.getClass().getName() + " factory";
    }


    public Widget mousePressed(MouseEvent e, History history, Lexicon lookupLexicon, Lexicon newLexicon) {

        System.out.println("FormFactoryController.mousePressed");
        
        Point p = e.getPoint();
        
        if( getArchetype().contains(p)) {
            Form f = createForm();
            newLexicon.put(new Name(), f);
            return f;
        }

        return null;
    }

    public boolean mouseClicked(MouseEvent e, History history, Lexicon lexicon)
    {
        return false;
    }
    
    public void release(Point p, History history, Lexicon lookupLexicon, Lexicon newlexicon)
    {
    }
    
    public void drag(Point p, History history, Lexicon lexicon)
    {
    }

    public Point getCenter()
    {
        return getArchetype().getCenter();
    }

    public void setHighlighted(boolean h)
    {
    }

    public Point getBounds()
    {
        return getArchetype().getBounds();
    }

}

File: ai/src/main/scala/org/pingel/gestalt/core/FormFactory.scala
package org.pingel.gestalt.core

import java.awt.Point
import java.awt.event.MouseEvent
import org.pingel.gestalt.ui.Widget

case class FormFactory(archetype: Form) extends Widget {
  
	// removed during scalification:
	// public FormFactory(){}

    def createForm(): Form = {
        // This is called by mousePressed
        var f = archetype.duplicate()
        f.arrange(getArchetype().getCenter())
        f.setDetachable(true)
        f
    }

//    public Form createForm(Unifier unifier)
//    {
//        // This is called by org.pingel.causality.docalculus.*
//        
//        Form f = archetype.duplicateAndReplace();
//        f.arrange(getArchetype().getCenter());
//        f.setDetachable(true);
//    		
//        // TODO bind vars in unifier
//        return null;
//    }

    def getArchetype() = archetype

    def toString() = archetype.getClass().getName() + " factory"

    def mousePressed(e: MouseEvent, history: History, lookupLexicon: Lexicon, newLexicon: Lexicon): Widget = {
        println("FormFactoryController.mousePressed")
        val p = e.getPoint()
        if( getArchetype().contains(p)) {
            val f = createForm()
            newLexicon.put(new Name(), f)
            return f
        }
        null
    }

    def mouseClicked(e: MouseEvent, history: History, lexicon: Lexicon) = false

    def release(p: Point, history: History, lookupLexicon: Lexicon, newlexicon: Lexicon): Unit = { }
    
    def drag(p: Point, history: History, lexicon: Lexicon): Unit = {}

    def getCenter() = getArchetype().getCenter()

    def setHighlighted(h: Boolean): Unit = { }

    def getBounds() = getArchetype().getBounds()

}
